### PR TITLE
memset/memcpy/memmove expansion to fix bug with select

### PR DESCRIPTION
**Pull request by azhogin**
Added pass to expand memset/memcpy/memmove intrinsics to fix problems with selects in sroa.

### CODE REVIEW

File: llvm/lib/Target/TVM/TVMLowerIntrinsics.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.addRequired<TargetTransformInfoWrapperPass>();
+  }
+};
+
+}
+
+char TVMLowerIntrinsics::ID = 0;
+
+//char &llvm::TVMLowerIntrinsicsID = TVMLowerIntrinsics::ID;
```
Comment (akiramenai) [inline llvm/lib/Target/TVM/TVMLowerIntrinsics.cpp:47]: Can we remove it?

File: llvm/lib/Target/TVM/TVMLowerIntrinsics.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    ++I;
+
+    switch (ID) {
+    case Intrinsic::memcpy: {
+      auto *Memcpy = cast<MemCpyInst>(Inst);
+      Function *ParentFunc = Memcpy->getParent()->getParent();
+      const TargetTransformInfo &TTI =
+          getAnalysis<TargetTransformInfoWrapperPass>().getTTI(*ParentFunc);
+      expandMemCpyAsLoop(Memcpy, TTI);
+      Changed = true;
```
Comment (akiramenai) [inline llvm/lib/Target/TVM/TVMLowerIntrinsics.cpp:67]: Perhaps we can `return false;` in the `default` clause and move common code (`return true` + `eraseFromParent` out of the switch).

File: llvm/lib/Target/TVM/TVMLowerIntrinsics.cpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+  for (Function &F : M) {
+    if (!F.isDeclaration())
+      continue;
+
+    switch (F.getIntrinsicID()) {
+    case Intrinsic::memcpy:
+    case Intrinsic::memmove:
+    case Intrinsic::memset:
+      if (expandMemIntrinsicUses(F))
```
Comment (akiramenai) [inline llvm/lib/Target/TVM/TVMLowerIntrinsics.cpp:103]: I doubt that we need to double-check intrinsic IDs. It also seems that `if` can't be `false`, can it?

Comment (azhogin) [reply llvm/lib/Target/TVM/TVMLowerIntrinsics.cpp:103]: First we enumerate all declared functions in the module. If function is a specific intrinsic, we enumerate all usages of it to expand. So, we need first check to decide make usages enumeration or skip the function.
`if` can be `false` when intrinsic is declared in module, but never called.

File: llvm/lib/Target/TVM/TVMLowerIntrinsics.cpp (hunk 2 above)
Comment (azhogin) [reply llvm/lib/Target/TVM/TVMLowerIntrinsics.cpp]: Fixed.

File: llvm/lib/Target/TVM/TVMLowerIntrinsics.cpp (hunk 1 above)
Comment (azhogin) [reply llvm/lib/Target/TVM/TVMLowerIntrinsics.cpp]: Fixed.

File: llvm/tools/clang/test/CodeGen/tvm/bug-select-with-memcpy.cpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
+// RUN: %clang -O3 -target tvm %s --sysroot=%S/../../../../../projects/ton-compiler/cpp-sdk/
```
Comment (akiramenai) [inline llvm/tools/clang/test/CodeGen/tvm/bug-select-with-memcpy.cpp:1]: It requires linker, could we add `-S` or move this test to the test suite?

Comment (azhogin) [reply llvm/tools/clang/test/CodeGen/tvm/bug-select-with-memcpy.cpp]: Fixed.

**Review by akiramenai — APPROVED** [review]

**Pull request merged**